src/database/pvRecord.cpp

```diff
@@ -77,22 +77,28 @@ void PVRecord::destroy()
         isDestroyed = true;
         pvTimeStamp.detach();
     
-        std::list<PVRecordClientPtr>::iterator clientIter;
+        std::list<PVRecordClientWPtr>::iterator clientIter;
         while(true) {
             clientIter = pvRecordClientList.begin();
             if(clientIter==pvRecordClientList.end()) break;
+            PVRecordClientPtr client = clientIter->lock();
+            if(client.get()) {
+                unlock();
+                client->detach(getPtrSelf());
+                lock();
+            }
             pvRecordClientList.erase(clientIter);
```

Comment:
The most obvious issue here is that explicit unlock()/lock() leaves open the possibility that an exception thrown be 'client->detach()' would no re-lock.  There are several ways to correct this.  One is explicit try/catch/rethrow.

```c++
if(client) {
  unlock();
  try {
    client->detach(getPtrSelf());
    lock();
  } catch(...) {
    lock();
    throw;
  }
}
```
I personally find this verbose and potentially error prone as the re-lock() call should be duplicated.  My preference is to add a "guard" class to handle unlock() and re-lock().  So a definition is added to some header.

```c++
template<typename T>
struct XUnLocker {
  T* ref;
  XUnLocker(T& r) :ref(&r) { r.unlock(); }
 ~XUnLocker() { r.lock(); }
};
```
Then the code in question becomes:

```c++
if(client) {
  XUnLocker<PVRecord> L(*this); // calls unlock()
  client->detach(getPtrSelf());
  // lock() when 'L' goes out of scope due to normal flow or exception unwind
}
```